modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordExecDialog.kt

```diff
@@ -1,234 +1,335 @@
-
-        val jbTargets = (targetsState.targets).asJBList()
-        val searchField = JTextField().apply {
-            val field = this
-
-            // Add an informative placeholder.
-            val previousForeground = foreground
-            text = searchPlaceHolder
-            foreground = JBColor.GRAY
-            addFocusListener(object : FocusListener {
-                override fun focusGained(e: FocusEvent) {
-                    if (field.text.equals(searchPlaceHolder)) {
-                        field.text = ""
-                        field.foreground = previousForeground
-                    }
-                }
+    private var refreshing: Boolean = false
 
-                override fun focusLost(e: FocusEvent) {
-                    if (field.text.isEmpty()) {
-                        field.foreground = JBColor.GRAY
-                        field.text = searchPlaceHolder
-                    }
-                }
-            })
-
-            // Add filtering logic on search field update.
-            document.addDocumentListener(object : DocumentListener {
-                override fun insertUpdate(e: DocumentEvent) = updateList()
-                override fun removeUpdate(e: DocumentEvent) = updateList()
-                override fun changedUpdate(e: DocumentEvent) = updateList()
-
-                private fun updateList() {
-                    val searchTerm = field.text
-                    if (!searchTerm.equals(searchPlaceHolder)) {
-                        targetsState.searchPhrase = searchTerm
-                        jbTargets.setListData(targetsState.targets.toTypedArray())
-                    }
-                }
-            })
+    /**
+     * List of targets available in the current namespace.
+     */
+    private val targetOptions: JBList<String> = JBList(emptyList<String>()).apply {
+        selectionMode = ListSelectionModel.SINGLE_SELECTION
+        minimumSize = Dimension(250, 350)
+    }
 
-            // Add focus logic so that the user can change back and forth from search field
-            // to target selection using tab/shift+tab.
-            addKeyListener(object : KeyListener {
-                override fun keyTyped(p0: KeyEvent) {
-                }
+    /**
+     * Dropdown allowing for switching namespaces.
+     */
+    private val namespaceOptions: ComboBox<String> = ComboBox(object : DefaultComboBoxModel<String>() {
+        override fun setSelectedItem(anObject: Any?) {
+            super.setSelectedItem(anObject)
 
-                override fun keyPressed(e: KeyEvent) {
-                    if (e.keyCode == KeyEvent.VK_TAB) {
-                        if (e.modifiersEx > 0) {
-                            field.transferFocusBackward()
-                        } else {
-                            field.transferFocus()
-                        }
-                        e.consume()
-                    }
-                }
+            if (refreshing) {
```

Comment:
I would hope there is some sort of "watcher" for this kind of thing, but if this works it's fine.

You don't deserve further suffering.